### PR TITLE
[SDK-686] idtype and request t

### PR DESCRIPTION
**Pull request by turtledreams**
checking bulk still, tests are working fine

### CODE REVIEW

**Comment by turtledreams** [comment]
Hmm there are some linting issues. Will fix

File: lib/countly-common.js [hunk 1]
```diff
@@ -23,6 +23,18 @@ var cc = {
         DEBUG: '[DEBUG] ',
         VERBOSE: '[VERBOSE] ',
     },
+    /**
+     * 
+     *device ID type:
+     *0 - device ID was set by the developer during init
+     *1 - device ID was auto generated by Countly
+     *2 - device ID was temporarily given by Countly
+     *3 - device ID was provided from location.search
```
Comment (ArtursKadikis) [inline lib/countly-common.js:32]: should not exist

File: test/tests_device_id_type.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    assert.ok(cc.isUUID(providedDeviceId));
+}
+function checkRequestsForT(queue, expectedInternalType) {
+    for (var i = 0; i < queue.length; i++) {
+        assert.ok(queue[i].t);
+        assert.equal(queue[i].t, expectedInternalType);
+    }
+}
+
+describe("View test", () => {
```
Comment (ArtursKadikis) [inline test/tests_device_id_type.js:27]: sloppy

File: test/tests_device_id_type.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            checkRequestsForT(rq, cc.deviceIdTypeEnums.SDK_GENERATED);
+            done();
+        }, hp.sWait);
+    });
+    it("Change generated device ID", (done) => {
+        // clear previous data
+        hp.clearStorage();
+        // initialize SDK
+        initMain(undefined);
+        Countly.change_id("changedID");
```
Comment (ArtursKadikis) [inline test/tests_device_id_type.js:48]: add type check before here

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Let's add the "second batch" of tests like in Web

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+## 22.02.0
+- Now you can get the device ID type of the user
```
Comment (ArtursKadikis) [inline CHANGELOG.md:2]: * "Added call to..."

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,7 @@
...
+- Now you can get the device ID of the user
```
Comment (ArtursKadikis) [inline CHANGELOG.md:3]: * "Added call to..."

File: lib/countly.js [hunk 6]
```diff
@@ -101,6 +102,8 @@ Countly.Bulk = Bulk;
  * @param {number} [conf.session_update=60] - how often in seconds should session be extended
  * @param {number} [conf.max_events=100] - maximum amount of events to send in one batch
  * @param {boolean} [conf.force_post=false] - force using post method for all requests
+ * @param {boolean} [conf.erase_stored_id=false] - set it to true if you want to erase the stored device ID
```
Comment (ArtursKadikis) [inline lib/countly.js:105]: this should be named correctly. "clearStoredDeviceID"

File: lib/countly.js [hunk 7]
```diff
@@ -72,6 +72,7 @@ Countly.Bulk = Bulk;
     var maxStackTraceLinesPerThread = 30;
     var maxStackTraceLineLength = 200;
     var __data = {};
+    var deviceIdType = cc.deviceIdTypeEnums.SDK_GENERATED;
```
Comment (ArtursKadikis) [inline lib/countly.js:75]: there shouldn't be any default type set

File: lib/countly.js [hunk 8]
```diff
@@ -251,6 +269,7 @@ Countly.Bulk = Bulk;
         maxStackTraceLinesPerThread = 30;
         maxStackTraceLineLength = 200;
         __data = {};
+        deviceIdType = cc.deviceIdTypeEnums.SDK_GENERATED;
```
Comment (ArtursKadikis) [inline lib/countly.js:272]: there shouldn't be any default type set

File: lib/countly.js [hunk 9]
```diff
@@ -179,15 +184,28 @@ Countly.Bulk = Bulk;
                 // eslint-disable-next-line no-console
                 console.log(ex.stack);
             }
+            if (conf.erase_stored_id) {
```
Comment (ArtursKadikis) [inline lib/countly.js:187]: add a comment

File: lib/countly.js [hunk 10]
```diff
@@ -179,15 +184,28 @@ Countly.Bulk = Bulk;
...
+                storeSet("cly_id", null);
+                storeSet("cly_id_type", null);
+            }
 
             if (Countly.url === "") {
                 cc.log(cc.logLevelEnums.ERROR, "Please provide server URL");
             }
             else {
                 cc.log(cc.logLevelEnums.INFO, "Countly initialized");
                 if (cluster.isMaster) {
```
Comment (ArtursKadikis) [inline lib/countly.js:198]: add comments of what is happening afterwards

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,7 @@
+## 22.02.0
```
Comment (ArtursKadikis) [inline CHANGELOG.md:1]: major breaking change should be mentioned

File: lib/countly.js [hunk 12]
```diff
@@ -179,15 +184,28 @@ Countly.Bulk = Bulk;
...
-                    Countly.device_id = conf.device_id || Countly.device_id || getId();
+                    var storedId = storeGet("cly_id", null);
```
Comment (ArtursKadikis) [inline lib/countly.js:199]: we should be getting the previous type

File: lib/countly.js [hunk 13]
```diff
@@ -1389,11 +1429,17 @@ Countly.Bulk = Bulk;
...
-    *  Get device ID, stored one, or generate new one
+    *  If an ID is provided sets the device ID  type, or generates new device ID
+    * @param {String} storedDeviceId - stored device ID to check
     *  @returns {String} device id
     */
-    function getId() {
-        return storeGet("cly_id", null) || cc.generateUUID();
+    function assignId(storedDeviceId) {
+        if (storedDeviceId) {
+            deviceIdType = storeGet("cly_id_type", null);
```
Comment (ArtursKadikis) [inline lib/countly.js:1438]: get this out

File: lib/countly.js [hunk 14]
```diff
@@ -1389,11 +1429,17 @@ Countly.Bulk = Bulk;
...
 
     /**
-    *  Get device ID, stored one, or generate new one
+    *  If an ID is provided sets the device ID  type, or generates new device ID
+    * @param {String} storedDeviceId - stored device ID to check
     *  @returns {String} device id
     */
-    function getId() {
-        return storeGet("cly_id", null) || cc.generateUUID();
+    function assignId(storedDeviceId) {
```
Comment (ArtursKadikis) [inline lib/countly.js:1436]: eliminate this function

File: lib/countly.js [hunk 15]
```diff
@@ -179,15 +184,28 @@ Countly.Bulk = Bulk;
...
                 cc.log(cc.logLevelEnums.INFO, "Countly initialized");
                 if (cluster.isMaster) {
-                    Countly.device_id = conf.device_id || Countly.device_id || getId();
+                    var storedId = storeGet("cly_id", null);
+                    if ((conf.device_id || Countly.device_id) && !storedId) {
+                        Countly.device_id = conf.device_id || Countly.device_id;
+                        deviceIdType = cc.deviceIdTypeEnums.DEVELOPER_SUPPLIED;
+                    }
+                    else {
+                        Countly.device_id = assignId(storedId);
```
Comment (ArtursKadikis) [inline lib/countly.js:205]: split this up into explicit calls

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Review by ArtursKadikis — APPROVED** [review]

**Pull request merged**